xla/service/gpu/cudnn_norm_rewriter.cc

```diff
@@ -254,43 +269,49 @@ class CudnnNormRewriterVisitor : public DfsHloRewriteVisitor {
     return MatchLayerNorm(instr);
   }
 
-  Status HandleDivide(HloInstruction* instr) override {
-    return MatchNormfactor(instr);
-  }
-
   // Matches and rewrites a layer norm pattern with scale multiplication and
-  // bias addition into a Custom Call to cuDNN. The pattern matching employs
-  // modularity to cover the scope of realizations.
+  // bias addition into a Custom Call to cuDNN.
   Status MatchLayerNorm(HloInstruction* instr) {
-    HloInstruction *input, *input0, *input1, *input2, *epsilon, *scale, *bias,
-        *expectation, *reduce;
+    HloInstruction *input, *input0, *input1, *input2, *scale, *bias, *epsilon,
+        *expectation, *expectation0, *reduce, *norm_factor, *variance;
     if (Match(instr,
               SubtractMultiplyAddAnyOrder(
                   m::Op(&input),
                   Expectation(&expectation, &reduce, m::Op(&input0)),
-                  NormFactor(&input1, &input2, &epsilon),
+                  NormFactor(&norm_factor, &input1, &input2, &variance,
+                             &expectation0, &epsilon),
+                  // NormFactor(&input1, &input2, &epsilon),
                   m::Broadcast(m::Op(&scale)), m::Broadcast(m::Op(&bias))))) {
 #if CUDNN_VERSION < 8905
-      // Layer Norm kernels are available with cuDNN 8.9.5 and above.
+      // Layer norm kernels are available with cuDNN 8.9.5 and above.
       VLOG(1) << "Layer norm Custom Calls require cuDNN 8.9.5.";
       return OkStatus();
 #endif  // CUDNN_VERSION < 8905
 
-      // Layer norm kernels require Ampere or newer architectures.
-      if (!cuda_compute_capability_.IsAtLeast(
-              se::CudaComputeCapability::AMPERE)) {
-        VLOG(1)
-            << "Layer norm Custom Calls require Ampere or newer architecture.";
+      if (!instr->GetModule()
+               ->config()
+               .debug_options()
+               .xla_gpu_enable_cudnn_layer_norm()) {
+        VLOG(1) << "Layer norm Custom Calls disabled.";
+        return OkStatus();
+      }
+
+      // Layer norm kernels require Ampere or Hopper architectures.
+      if (!cuda_compute_capability_.Is(se::CudaComputeCapability::AMPERE) &&
+          !cuda_compute_capability_.Is(se::CudaComputeCapability::HOPPER)) {
+        VLOG(1) << "Layer norm Custom Calls require Ampere or Hopper "
+                   "architectures.";
         return OkStatus();
       }
 
       // Verify the uniqueness of the inputs.
       auto is_input = [input](HloInstruction* inputx) -> bool {
         return inputx->unique_id() == input->unique_id() ||
-               inputx->operand_count() == 1 &&
+               inputx->opcode() == HloOpcode::kConvert &&
```

Comment:
I think we don’t need to verify the convert here since we verify later that input has a supported type. If the convert is to an unsupported type, then somewhere in the graph there would have to be a conversion from that unsupported type to a supported type.